inc/azure_macro_utils/macro_utils.h

```diff
@@ -322,6 +322,26 @@ __pragma(warning(pop))
 #endif
 
 
+/*the below 2 macros do the same thing: they take a construct in the form of ORIGINAL_NAME(ARG1, ARG2,...) and expand to 
+* SUBST_NAME(ARG1, ARG2,...), so effectively it substitutes one token with another. Here's some example:
+* MU_SUBSTITUTE(SUM(7,3), DIF) instead of doing something with SUM(7,3) will expand to DIFF(7,3). In order for this to work, the user needs to mark "SUM" as a replacebable macro by
+* introducing the below define:
+* #define MU_IS_SUBST_SUM
+* 
+* L1, L2... are needed because of C preprocessor's expansion rules (a macro expansion cannot re-expand the same name). So user be aware!
```

Comment:
>L1, L2... are needed because of C preprocessor's expansion rules (a macro expansion cannot re-expand the same name). So user be aware! [](start = 2, length = 134)

I don't exactly understand why we have L1,L2,L3. Does that mean each compilation unit needs to only have one instance of `MU_SUBSTITUTE_CALL_Lx`? Or one instance per ORIG_MACRO? Or some other limitation?

Reply:
I will try to explain why we have different names.

Assume you have the following macro:
```c
#define F(X) F(X+1)```
```

or the other following 2 macros:
```c
#define M1(X) M2(X)+1
#define M2(X) M1(X) +2
```

Note how these macros are recursive: F is recursive with itself, while M1 and M2 are indirectly recursive. The preprocessor will not infinitely expand them, the preprocessor will actually stop expanding macros that it already expanded and leave the substituted tokens in the state that was there when the recursivity was detected. This in itself is a trick (and we don't use it anywhere in our code AFAIK).

So it might be that across the expansion of some macro, MU_SUBSTITUTE was already expanded and then it would not be re-expanded.

To answer your questions:
1) An translation unit can have as many of the MU_SUBSTITUTE as it wants, as long as they do not cause MU_SUBST to be re-expanded
2) You can apply it to as many of the macros you want...
3) I am not aware of limitations, except the ones which require a different macro name for the same purpose (avoid re-expansion of the same macro).


---
In reply to: [507999029](https://github.com/Azure/azure-macro-utils-c/pull/49#discussion_r507999029) [](ancestors = 507999029)